src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/TeredoHelper.cs

```diff
@@ -0,0 +1,165 @@
+
+                    // Async completion: add us to the s_pendingNotifications list so we'll be canceled in the
+                    // event of an AppDomain unload.
+                    s_pendingNotifications.Add(helper);
+                    return false;
+                }
+            }
+
+            if (err != Interop.IpHlpApi.ERROR_SUCCESS)
+            {
+                throw new Win32Exception((int)err);
+            }
+
+            return true;
+        }
+
+        private void RunCallback(object o)
+        {
+            GlobalLog.Assert(_runCallbackCalled, "RunCallback called without setting runCallbackCalled!");
+
+            // If OnAppDomainUnload beats us to the lock, do nothing: the AppDomain is going down soon anyways.
+            // Otherwise, wait until the call to CancelMibChangeNotify2 is done before giving it up.
+            lock (s_pendingNotifications)
+            {
+                if (Environment.HasShutdownStarted)
+                {
+                    return;
+                }
+
+#if DEBUG
+                bool successfullyRemoved = s_pendingNotifications.Remove(this);
+                GlobalLog.Assert(successfullyRemoved,
+                    "RunCallback for a TeredoHelper which is not in s_pendingNotifications!");
+#else
+                s_pendingNotifications.Remove(this);
+#endif
+
+                GlobalLog.Assert(_cancelHandle != null && !_cancelHandle.IsInvalid,
+                    "Invalid cancelHandle in RunCallback");
+
+                _cancelHandle.Dispose();
+            }
+
+            _callback.Invoke(_state);
+        }
+
+        // This callback gets run on a native worker thread, which we don't want to allow arbitrary user code to
+        // execute on (it will block AppDomain unload, for one).  Free the MibTable and delegate (exactly once)
+        // to the managed ThreadPool for the rest of the processing.
+        //
+        // We can't use SafeHandle here for table because the marshaller doesn't support them in reverse p/invokes.
+        // We won't get an AppDomain unload here anyways, because OnAppDomainUnloaded will block until all of these
+        // callbacks are done.
+        private void OnStabilized(IntPtr context, IntPtr table)
+        {
+            Interop.IpHlpApi.FreeMibTable(table);
+
+            // Lock the TeredoHelper instance to ensure that only the first call to OnStabilized will get to call 
+            // RunCallback.  This is the only place that TeredoHelpers get locked, as individual instances are not
+            // exposed to higher layers, so there's no chance for deadlock.
+            if (!_runCallbackCalled)
+            {
+                lock (this)
+                {
+                    if (!_runCallbackCalled)
+                    {
+                        _runCallbackCalled = true;
+                        ThreadPool.QueueUserWorkItem(RunCallback, null);
+                    }
+                }
+            }
+        }
+    }
+}
```

Comment:
Very odd that for me GitHub doesn't show this file lacking the new-line at EOF.
Fixed.